src/compiler/crystal/semantic/type_lookup.cr

```diff
@@ -279,6 +264,41 @@ class Crystal::Type
         type_vars << type.virtual_type
       end
 
+      splat_index = instance_type.splat_index
+      var_count = instance_type.type_vars.size
+
+      case {type_vars.any?(TypeSplat), splat_index}
+      in {true, Int32}
+        # node is `(A, B, C, *D, E)`, definition is `(T, *U, V, W)`; *D would splat into V
```

Comment:
Could you explain this a bit more? Aren't all types on the left hand side need to be know to know how to splat them? I don't understand why D would splat only into V, which is not a splat.

Reply:
The arguments before and after the splat in the instantiation are always matched first, so `A`, `B`, and `C` would match `T` and `*U`, followed by `E` with `W`, and finally `*D` with `*U` and `V`, which would be disallowed regardless of what types the `TypeParameter`s and `TypeSplat`s have. Thus the definition must not have more than 3 non-splat parameters before `*U` or more than 1 non-splat parameter after `*U`.

Reply:
In the PR description it says:

```crystal
class Foo(T, *U, V); end
class Bar(*A) < Foo(*A, Int32, Int32); end        # error, *A splats into T
```

But if I do:

```crystal
Bar(Int32, Int32, Int32).new
```

then we get:

```crystal
Foo(Int32, Int32, Int32, Int32, Int32)
```

and for:

```crystal
Foo(T, *U, V)
```

we get:

- T: Int32
- U: Int32, Int32, Int32
- V: Int32

So that shouldn't be an error?

That said, I'm sure I'm misunderstanding something.

Reply:
I guess what I'm saying is... don't we need to expand/collect all the types on `node` and then match that against the definition?

Reply:
In light of https://github.com/crystal-lang/crystal/issues/3649#issuecomment-864298739, maybe this restriction isn't really necessary...